.github/workflows/test.yml

```diff
@@ -122,10 +122,10 @@ jobs:
         run: cargo test
```

Comment:
One note to continue the performance thread. We already have a few benchmarks. It would probably be a good idea to add quick tests for parsing a few different styles of mailbox and mailboxes and comparing before and after. I agree that the maintainability and correctness of using a proper parsing library is almost certainly worth it but it would be nice to have some rough numbers just to understand the difference.

Reply:
Before:

```
     Running benches/mailbox_parsing.rs (target/release/deps/mailbox_parsing-7dafc3dc7f99af43)
parse single mailbox    time:   [103.40 ns 103.67 ns 103.96 ns]

parse multiple mailboxes
                        time:   [108.19 ns 108.34 ns 108.50 ns]
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  3 (3.00%) high severe
```

After:

```
     Running benches/mailbox_parsing.rs (target/release/deps/mailbox_parsing-382c53ca32910006)
parse single mailbox    time:   [63.761 µs 63.879 µs 64.022 µs]
                        change: [+61383% +61572% +61766%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  6 (6.00%) high severe

parse multiple mailboxes
                        time:   [61.371 µs 61.531 µs 61.704 µs]
                        change: [+55475% +56118% +56517%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
```

I did not expect such a scary difference…

Reply:
Yeah, I would expect the cost to be dominated by the string allocations, so it would be mostly similar in cost. Although honestly 60 µs may still be acceptable.

One idea may be trying to operate on `u8` rather than `char`. I think the only hard part is ensuring that you don't splice characters. You may also end up paying for UTF-8 validation on reconstruction which may negate the wins.

Or just try simplifying the parser and see if some simplification cuts the cost dramatically. Maybe it is just one subparser that costs. I'll try to look into it if I have time tonight.

Reply:
I redid with the right benchmarks and with consuming all the data till the end. Before:

```
     Running benches/mailbox_parsing.rs (target/release/deps/mailbox_parsing-7dafc3dc7f99af43)
parse single mailbox    time:   [115.99 ns 116.27 ns 116.56 ns]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

parse multiple mailboxes
                        time:   [514.04 ns 514.82 ns 515.58 ns]
Found 21 outliers among 100 measurements (21.00%)
  2 (2.00%) low mild
  19 (19.00%) high mild
```

After:

```
     Running benches/mailbox_parsing.rs (target/release/deps/mailbox_parsing-382c53ca32910006)
parse single mailbox    time:   [83.718 µs 83.878 µs 84.108 µs]
                        change: [+71896% +72142% +72401%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 24 outliers among 100 measurements (24.00%)
  24 (24.00%) high severe

parse multiple mailboxes
                        time:   [204.38 µs 204.60 µs 204.83 µs]
                        change: [+39355% +39469% +39579%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low severe
  3 (3.00%) low mild
  3 (3.00%) high mild
```

> I'll try to look into it if I have time tonight.

Alright, I will try also from my side.

Reply:
I tired caching the parser creation for mailbox_list and it only saves a few percentage points.

Reply:
I poked around and saw two main sources of slowness.

1. Lots of rules return a `Vec<char>` this is allocating a lot of `Vec`s. Making some simple changes such as adding special rules for skipping whitespace without allocating a `Vec` and making `quoted_string` avoid `Vec` for it's components seems to be able to half the runtime.
2. Failed parses seem slow. By reordering `choice`s to match the test case in the benchmark I can save 1/4 of the runtime. This obviously depends on the input data but may show us where a lot of the cost is. I don't know if there is a way to avoid this by avoiding generating errors for the initial members of a `choice` or somehow making errors cheaper.

Reply:
1. Not sure to understand your point. What do you mean by adding special rules? Half of parsers return a `Vec<char>`, I am not sure to see how to remove them.
2. I replaced the `Simple` chumsky error manager by the `Cheap` one and it seems to have a significative (positive) impact:

```
     Running benches/mailbox_parsing.rs (target/release/deps/mailbox_parsing-382c53ca32910006)
parse single mailbox    time:   [26.021 µs 26.047 µs 26.077 µs]
                        change: [-66.298% -66.211% -66.124%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

parse multiple mailboxes
                        time:   [51.289 µs 51.336 µs 51.380 µs]
                        change: [-72.143% -72.099% -72.056%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
```

Reply:
> What do you mean by adding special rules?

For example I made a `skip_fws()` parser that returned `()` and didn't allocate any `Vec` inside of it. I did this for a bunch of the rules where the result is usually ignored.

> Half of parsers return a Vec<char>, I am not sure to see how to remove them.

I tried to make a single point where we switched from `char` to `Vec<char>` (should probably actually be String) on each "path". So for example I switched the sub-parsers used in `quoted_string` to all return `char` and the collection to `Vec<char>` was done only once inside of that parser. This avoids allocating tiny `Vec` for each sub `fws`, This can also be an issue in `phrase` but I doubt it is a hot path, most emails just have one `quoted-string` for the display name.

Reply:
> For example I made a `skip_fws()` parser that returned `()` and didn't allocate any `Vec` inside of it. I did this for a bunch of the rules where the result is usually ignored.

Oh I see, indeed it could improve perfs. Would you like to push on the branch?

Reply:
No it's fine, I have some free time right so I will try to do those changes :slightly_smiling_face:

Reply:
I think it is an acceptable cost to have a proper parser. Paolo will have the final call though.